client/app/pages/queries/hooks/useQueryExecute.js

```diff
@@ -1,79 +1,120 @@
+      }
 
-  const queryResultData = useQueryResult(queryResult);
-  const isQueryExecuting = useMemo(() => !!queryResult && !includes(["done", "failed"], queryResultData.status), [
-    queryResult,
-    queryResultData.status,
-  ]);
+      queryResultInExecution.current = newQueryResult;
 
-  const [isExecutionCancelling, setIsExecutionCancelling] = useState(false);
+      setExecutionState({
+        updatedAt: newQueryResult.getUpdatedAt(),
+        executionStatus: newQueryResult.getStatus(),
+        isExecuting: true,
+        cancelCallback: () => {
+          recordEvent("cancel_execute", "query", query.id);
+          setExecutionState({ isCancelling: true });
+          newQueryResult.cancelExecution();
+        },
+      });
 
-  const executeQuery = useCallback(() => {
-    recordEvent("execute", "query", query.id);
-    setQueryResult(query.getQueryResult(0));
-  }, [query]);
+      const onStatusChange = status => {
+        if (queryResultInExecution.current === newQueryResult) {
+          setExecutionState({ updatedAt: newQueryResult.getUpdatedAt(), executionStatus: status });
+        }
+      };
 
-  const executeAdhocQuery = useCallback(
-    selectedQueryText => {
-      recordEvent("execute", "query", query.id);
-      setQueryResult(query.getQueryResultByText(0, selectedQueryText));
+      newQueryResult
+        .toPromise(onStatusChange)
+        .then(queryResult => {
+          if (queryResultInExecution.current === newQueryResult) {
+            // TODO: this should probably belong in the QueryEditor page.
+            if (queryResult && queryResult.query_result.query === query.query) {
+              query.latest_query_data_id = queryResult.getId();
+              query.queryResult = queryResult;
+            }
+
+            setExecutionState({
+              queryResult,
+              loadedInitialResults: true,
+              error: null,
+              isExecuting: false,
+              isCancelling: false,
+              executionStatus: null,
+            });
+          }
+        })
+        .catch(queryResult => {
+          if (queryResultInExecution.current === newQueryResult) {
+            setExecutionState({
+              queryResult,
+              loadedInitialResults: true,
+              error: queryResult.getError(),
+              isExecuting: false,
+              isCancelling: false,
+              executionStatus: ExecutionStatus.FAILED,
+            });
+          }
+        });
     },
     [query]
   );
 
-  const cancelExecution = useCallback(() => {
-    if (queryResult) {
-      recordEvent("cancel_execute", "query", query.id);
-      queryResult.cancelExecution();
-      setIsExecutionCancelling(true);
-    }
-  }, [query.id, queryResult]);
+  const queryRef = useRef(query);
+  const executeQueryRef = useRef(executeQuery);
+  useEffect(() => {
+    queryRef.current = query;
+    executeQueryRef.current = executeQuery;
+  }, [executeQuery, query]);
```

Comment:
This is the safest way I could think to remove the dependencies from the `useEffect` hook and make it only execute once.

Reply:
Is `useEffect` really needed here? Seems you can leave only assignment - assigning value to `ref.current` does not re-render component.